### PR TITLE
[WIP] Add Prometheus ideas

### PR DESCRIPTION
**Pull request by Gouthamve**
Super early draft, not ready for a merge.

I have added some mentors but not sure if you are up for it though and I am pretty sure I missed some names where people are interested. Please point out and I'll make the changes.

@brancz @superq @brian-brazil @tomwilkie @juliusv @fabxc

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -36,12 +36,96 @@ TODO
...
 
+Prometheus ideas:
+
+#### Streaming APIs
+* Description: Currently the HTTP API sends all the data in one go and that has the chance overwhelm prometheus. To guard against this, there are limits set in Prometheus. Make APIs streaming where data is retrieved and sent in chunks rather than at once.
+* Recommended Skills: golang
+* Mentor(s): Frederic Branczyk (@brancz), Goutham V (@gouthamve)
+* Issue: https://github.com/prometheus/prometheus/issues/3690
+
 #### Add option to log slow queries and recording rules
```
Comment (brian-brazil) [inline README.md]: 2.1 has recording rule eval time in the UI

**Comment by brancz** [comment]
This looks great! Thanks for putting this together @Gouthamve!

**Comment by Gouthamve** [comment]
@brian-brazil Can you suggest the order? Also, @RichiH I added you as a mentor to Outalator.

File: README.md [hunk 2]
```diff
@@ -36,12 +36,96 @@ TODO
...
+#### Features that aid in building and testing alerting expressions
+* Description: One of the most annoying (but important) things when using Prometheus is making sure that your alerting expressions are correct (semantically, syntactically, and referring to existing template variables). It would be great to be able to assist the user with that. This would be both frontend and backend / tooling work.
+* Recommended Skills: golang, javascript
+* Mentor(s): Julius Volz (@juliusv)
+* Issue: https://github.com/prometheus/prometheus/issues/1154 https://github.com/prometheus/prometheus/issues/1695 https://github.com/prometheus/prometheus/issues/1220 https://github.com/prometheus/prometheus/issues/1219
+
+#### Tracing in Prometheus and Alertmanager
+* Description: We can use open-tracing to trace several things in both prometheus and alertmanager, for example, prometheus query tracing.
+* Recommended Skills: golang
+* Mentor(s): Frederic Branczyk (@brancz), Tom Wilkie (@tomwilkie)
```
Comment (tomwilkie) [inline README.md]: +1

File: README.md [hunk 3]
```diff
@@ -36,12 +36,96 @@ TODO
...
+* Issue: https://github.com/prometheus/prometheus/issues/3689
+
+#### Features that aid in building and testing alerting expressions
+* Description: One of the most annoying (but important) things when using Prometheus is making sure that your alerting expressions are correct (semantically, syntactically, and referring to existing template variables). It would be great to be able to assist the user with that. This would be both frontend and backend / tooling work.
+* Recommended Skills: golang, javascript
+* Mentor(s): Julius Volz (@juliusv)
+* Issue: https://github.com/prometheus/prometheus/issues/1154 https://github.com/prometheus/prometheus/issues/1695 https://github.com/prometheus/prometheus/issues/1220 https://github.com/prometheus/prometheus/issues/1219
+
+#### Tracing in Prometheus and Alertmanager
+* Description: We can use open-tracing to trace several things in both prometheus and alertmanager, for example, prometheus query tracing.
```
Comment (tomwilkie) [inline README.md]: IMO this is related to "Add option to log slow queries", as one method for debugging slow queries could be to use the embedded tracer.

I would change the description to:
"We have some rudimentary OpenTracing instrumentation in Prometheus; this project would be to extend this and develop a basic built-in tracing implementation capable of debugging slow running queries.  In addition, we would like to extend OpenTracing instrumentation to Alertmanager, and correlate traces between the two systems."

@brancz @Gouthamve WDYT?

File: README.md [hunk 4]
```diff
@@ -36,12 +36,96 @@ TODO
 
 Prometheus is an open-source systems monitoring and alerting toolkit: https://prometheus.io/
 
-#### Add option to log slow queries and recording rules
-* Description: Having something like PostgreSQL's log_min_duration_statement would be useful to debug performance problems. It would be great to collect detailed query information, like how many chunks were necessary to compute the result and how many had to be loaded from disk.
+Prometheus ideas:
+
+#### Streaming APIs
+* Description: Currently the HTTP API sends all the data in one go and that has the chance overwhelm prometheus. To guard against this, there are limits set in Prometheus. Make APIs streaming where data is retrieved and sent in chunks rather than at once.
```
Comment (tomwilkie) [inline README.md]: I'd be interested in this wrt the remote APIs - the remote read API really needs extending to be streaming, and I'd love to make it use gRPC client-side push.  Do you think this is related enough to be part of the same project, or a different project?

Comment (brancz) [reply README.md]: I think this is sort of an umbrella project for those efforts, and remote APIs are a part of that, but could be implemented incrementally.

File: README.md (hunk 3 above)
Comment (brancz) [reply README.md]: I think there is overlap, but also think that people probably want both slow query logging and tracing.

**Comment by caniszczyk** [comment]
going to merge this in, if you have updates, please send a new PR, thanks @Gouthamve

**Pull request merged**